### PR TITLE
feat(cast): add mnemonic key generation to wallet new

### PR DESCRIPTION
**Pull request by mitche50**
## Motivation

Generating new wallets only allows for a user to generate a private key and associated address.  Many wallets default import mnemonic seeds, so we should have an option to generate a mnemonic along with the wallet.

## Solution

I've added an `--mnemonic` flag to the `cast wallet new` command which when provided will generate and provide the mnemonic seed.  This uses the `tiny-bip39` (https://docs.rs/crate/tiny-bip39/latest) crate to generate a random 12 word mnemonic and feeds into the ethers `MnemonicBuilder`.

### CODE REVIEW

**Comment by roynalnaruto** [comment]
`ethers-signers` offers a mnemonic wallet, which provides a nice interface to work with and a `mnemonic_to_wallet` fallible method to convert to a `Wallet<SigningKey>>` type. My suggestion is that using this would simplify your changes and make it easier to maintain :)

**Comment by mitche50** [comment]
> `ethers-signers` offers a [mnemonic wallet](https://github.com/gakonst/ethers-rs/blob/master/ethers-signers/src/wallet/mnemonic.rs), which provides a nice interface to work with and a `mnemonic_to_wallet` fallible method to convert to a `Wallet<SigningKey>>` type. My suggestion is that using this would simplify your changes and make it easier to maintain :)

The problem with the `ethers-signers` library is it doesn't surface the mnemonic phrase used to generate the wallet in `build_random` and requires a phrase to be set in `build`.  The point of this PR was to provide the mnemonic in the return to the user so they can use this to store their seed vs. the hex private key only.

File: cli/src/cmd/cast/wallet/mod.rs [hunk 1]
```diff
@@ -77,10 +87,39 @@ pub enum WalletSubcommands {
...
+                        eprintln!("`{}` is not a valid mnemonic length: {:?}", word_count, word_counts);
+                        std::process::exit(1);
+                    }
+                    let mnemonic_type_result = MnemonicType::for_word_count(word_count);
+
+                    let mnemonic_type = match mnemonic_type_result {
+                        Ok(m_type) => m_type,
+                        Err(err) =>  {
+                            eprintln!("Error determining mnemonic type :( \n  {}", err);
+                            std::process::exit(1);
```
Comment (DaniPopes) [inline cli/src/cmd/cast/wallet/mod.rs:104]: Please avoid using process::exit, return instead with `eyre::bail!("msg")` or `return Err(eyre::eyre!("msg"))`

File: cli/src/cmd/cast/wallet/mod.rs [hunk 2]
```diff
@@ -77,10 +87,39 @@ pub enum WalletSubcommands {
 impl WalletSubcommands {
     pub async fn run(self) -> eyre::Result<()> {
         match self {
-            WalletSubcommands::New { path, unsafe_password, .. } => {
+            WalletSubcommands::New { path, unsafe_password, mnemonic, word_count, .. } => {
                 let mut rng = thread_rng();
+                if mnemonic {
+                    let word_counts = [12, 15, 18, 21, 24];
+                    if !word_counts.contains(&word_count) {
```
Comment (DaniPopes) [inline cli/src/cmd/cast/wallet/mod.rs:94]: I don't think this is necessary to check ourselves as Mnemonic will just throw an error

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Review by DaniPopes — APPROVED** [review]

**Comment by gakonst** [comment]
> The problem with the ethers-signers library is it doesn't surface the mnemonic phrase used to generate the wallet in build_random and requires a phrase to be set in build. The point of this PR was to provide the mnemonic in the return to the user so they can use this to store their seed vs. the hex private key only.

Would you be open to providing a PR to ethers-signers which adds these features?

Thanks for taking this on!

**Review by to4kag — APPROVED** [review]

**Comment by shidaxi** [comment]
Hey @mitche50 how is this feature going, could you merge to master?

**Comment by mitche50** [comment]
> Hey @mitche50 how is this feature going, could you merge to master?

The feature is functional when testing locally, but not sure if the team wants to merge this to master.  @gakonst requested this be a PR to ethers-signers vs. to foundry but I haven't had time to move it over.

**Comment by gakonst** [comment]
There's some WIP https://github.com/gakonst/ethers-rs/pull/2238 with comments from Matt in there!

**Comment by Evalir** [comment]
Closing as stale.

**Pull request closed without merging**